### ISSUE TITLE
Add Smart Importer Support

### ISSUE THREAD

**Issue by jason-riddle** [jason-riddle/ledger-main#13]
Read

- https://github.com/beancount/beangulp/blob/ead8a2517d4f34c7ac7d48e4ef6d21a88be7363c/examples/import.py#L50

- https://raw.githubusercontent.com/beancount/smart_importer/refs/heads/main/README.rst

- https://raw.githubusercontent.com/beancount/beangulp/refs/heads/master/examples/importers/csvbank.py

- https://raw.githubusercontent.com/beancount/beangulp/refs/heads/master/examples/importers/ofx.py

Generate the min amount of python code so that Beangulp is working/integrated properly with smart_inporter. The goal is to activate it with a cli flag. if smart importer is enabled, the goal is to have it predict the other leg of the transaction and/or the payee (add 2 CLI flags I guess), if either flag is used, then trying to fill in the payee/other leg with hard coded accounts is ignored.

Determine (based on how the library works and using your judgement and reasoning) whether this needs to integrate to work with csv, xml, ofx, etc file types OR should it instead just work with JSONL file types. I need you to confirm based on how it works, validate it works, and explain your reasoning.

Also determine where this new code needs to go, give reasoning.

also, read below for more info on smart_importer, ignore bean-extract stuff , that was hallucinated

---

# smart_importer for Beancount: Detailed Guide

**smart_importer** is a Python library that **augments Beancount importers with machine learning** to **predict missing transaction fields** (particularly counter-account postings and normalized payee names) based on historical Beancount data. It is *not* a standalone CSV parser but a layer on top of importers that supplies semantic classification. ([[GitHub](https://github.com/beancount/smart_importer?utm_source=chatgpt.com)][1])

---

## 1. Core Purpose

Beancount importers (e.g., for CSV, OFX, QIF) typically parse external data into **partial transactions** — they extract the date, payee text, and one leg (source account and amount), leaving the counterpart account unspecified. smart_importer uses **machine learning** to *predict* those missing fields using patterns learned from your existing historical ledger. ([[GitHub](https://github.com/beancount/smart_importer?utm_source=chatgpt.com)][1])

### What it Predicts

| Field    | Description                                                                        |
| -------- | ---------------------------------------------------------------------------------- |
| Postings | The other leg of a transaction (e.g., assign expense category account)             |
| Payees   | Normalized payee names to reduce variation (e.g., “STARBUCKS STORE” → “Starbucks”) |

The ML models work on text and metadata from previously entered transactions to predict targets on new imported data. ([[GitHub](https://github.com/beancount/smart_importer?utm_source=chatgpt.com)][1])

---

## 2. High-Level Architecture

smart_importer integrates with Beancount by wrapping existing importer objects (e.g., a CSV or custom importer). It acts **after the raw parser** has produced partial transactions:

```
External file → Importer (parse) → Partial Transactions → smart_importer (ML prediction) → Completed Transactions
```

It uses a hook or wrapper pattern:

* **Hooks**: Apply globally to all importers via Beangulp’s hook system.
* **Wrapper**: Wrap a specific importer instance. ([[GitHub](https://github.com/beancount/smart_importer?utm_source=chatgpt.com)][1])

smart_importer only processes **single-leg postings** missing counterpart accounts. It *does not modify* fully specified transactions. ([[GitHub](https://github.com/beancount/smart_importer?utm_source=chatgpt.com)][1])

---

## 3. Supported Data Inputs

smart_importer **does not parse file formats itself**. Supported formats depend on the underlying importer. Typical formats used in Beancount workflows include:

* **CSV**: Most common exported format from banks and credit card providers.
* **OFX / QFX**: Structured financial interchange formats supported by Beancount’s own OFX parser.
* **QIF**: Older financial interchange format consumed by community importers.
* **Custom API outputs / JSON**: When converter scripts (e.g., for bank APIs like Starling) produce Beancount entries. ([[beancount](https://beancount.github.io/docs/importing_external_data.html?utm_source=chatgpt.com)][2])

Any file type that a Beancount importer can transform into partial transactions is effectively supported by smart_importer when layered on top. ([[beancount](https://beancount.github.io/docs/importing_external_data.html?utm_source=chatgpt.com)][2])

---

### Installation

Install from PyPI:

```
pip install smart-importer
```

Dependencies include `scikit-learn`, `beancount`, `numpy`, and `scipy`. ([[Google Groups](https://groups.google.com/g/beancount/c/rjrbf6Y39ew?utm_source=chatgpt.com)][3])

#### As a Beangulp Hook

```python
from your_custom_importer import MyBankImporter
from smart_importer import PredictPayees, PredictPostings

CONFIG = [
    MyBankImporter('whatever', 'config', 'is', 'needed'),
]

HOOKS = [
    PredictPostings().hook,
    PredictPayees().hook
]
```

This enables smart_importer for all importers in `CONFIG`. ([[GitHub](https://github.com/beancount/smart_importer?utm_source=chatgpt.com)][1])

#### Wrapping a Single Importer

```python
from smart_importer import PredictPayees, PredictPostings

CONFIG = [
    PredictPostings().wrap(
        PredictPayees().wrap(
            MyBankImporter('whatever', 'config', 'is', 'needed')        
        )
    ),
]
```

This creates a predictor-enhanced importer just for a specific source. ([[GitHub](https://github.com/beancount/smart_importer?utm_source=chatgpt.com)][1])

---

## 5. Training Data

smart_importer requires **existing transactions** as training data. It learns patterns of payee text and past posting accounts.

### Supplying Training Data

* With Fava (a Beancount web UI), **training data is automatically taken** from your loaded ledger.
* With `bean-extract`, include an existing ledger file using the `-e` option:

```
bean-extract -e existing_transactions.beancount import-config.py statement.csv
```

If training data is absent, smart_importer will report an error about empty training data. ([[GitHub](https://github.com/beancount/smart_importer?utm_source=chatgpt.com)][1])

---

## 6. How It Actually Works (Internals)

While the library does not explicitly document its internal ML model, the smart_importer implementation typically uses Python’s scikit-learn to train classifiers on historical text and posting patterns. The prediction pipeline uses tokenization of payee text and feature extraction to infer most likely account assignments and normalized payees based on past examples.

Tokenizers can be customized to improve text handling, e.g., using jieba for Chinese text segmentation. ([[GitHub](https://github.com/beancount/smart_importer?utm_source=chatgpt.com)][1])

---

## 7. Example End-to-End Workflow

**1. Write or choose a native importer:**
Example: `CSVImporter` for your bank’s CSV. ([[GitHub](https://github.com/beancount/smart_importer?utm_source=chatgpt.com)][1])

**2. Wrap with smart_importer:**
Add `PredictPostings` and `PredictPayees` around it. ([[GitHub](https://github.com/beancount/smart_importer?utm_source=chatgpt.com)][1])

**3. Run bean-extract with training data:**

```
bean-extract -e main.beancount import_config.py downloads/*.csv
```

**4. Review output:**
Generated `.beancount` file includes predicted postings and normalized payees. Edit if needed. ([[Mustachian Post Community](https://forum.mustachianpost.com/t/beancount-advancement-thread/2786?page=7&utm_source=chatgpt.com)][4])

**5. Merge into main ledger:**
Once validated, append to your primary ledger file.

---

### A. Confirm Training Data Is Loaded

Run `bean-extract` with the `-e` flag and inspect output for messages about training (errors appear if training data is empty). ([[Mustachian Post Community](https://forum.mustachianpost.com/t/beancount-advancement-thread/2786?page=7&utm_source=chatgpt.com)][4])

### B. Use Logging for Debugging

smart_importer uses Python’s standard logging. To see more detail:

```python
import logging
logging.getLogger('smart_importer').setLevel(logging.DEBUG)
```

Watch the logs during import to observe prediction actions. ([[CSDN Blog](https://blog.csdn.net/gitblog_00015/article/details/141744523?utm_source=chatgpt.com)][5])

### C. Compare Predictions to Ground Truth

1. Run importer without smart_importer.
2. Run it again with smart_importer.
3. Compare resulting `.beancount` files to ensure predicted postings/payees are present and sensible.

Manual verification and corrections help improve future predictions.

---

### Provide More and Better Training Data

Accuracy depends on patterns in your existing ledger. More labeled examples per payee and account increase classifier quality.

### Normalize Payee Text

Use consistent payee text in your historical entries to reduce noise and improve classification.

### Custom Tokenizers

For non-English text or domain-specific text, pass a custom tokenizer (e.g., jieba for Chinese) to `PredictPostings` to improve text feature extraction. ([[GitHub](https://github.com/beancount/smart_importer?utm_source=chatgpt.com)][1])

### Clean Importer Output

Ensure your importer produces **consistent partial transactions** with clean textual payee fields — this makes patterns easier to learn.

---

## 10. Known Limitations

* It may not perform well with **insufficient training data** — fallback is no prediction. ([[Mustachian Post Community](https://forum.mustachianpost.com/t/beancount-advancement-thread/2786?page=7&utm_source=chatgpt.com)][4])
* smart_importer only handles **single-leg postings** missing counterpart accounts; it doesn’t edit fully specified postings. ([[GitHub](https://github.com/beancount/smart_importer?utm_source=chatgpt.com)][1])
* Behavior may vary with differing formatting of CSV or other formats; some custom importers may require adjusting the extract interface. ([[Mustachian Post Community](https://forum.mustachianpost.com/t/beancount-advancement-thread/2786?page=7&utm_source=chatgpt.com)][4])

---

## 11. Related Tools and Ecosystem

There are third-party tools in the Beancount ecosystem that take inspiration from smart_importer, such as **beancount-categorizer** for regex-based automatic categorization, and **beancount-import** (web UI tool) that integrates classification and merging. ([[GitHub](https://github.com/bratekarate/beancount-categorizer?utm_source=chatgpt.com)][6])

---

## 12. References

* **smart_importer README** — core documentation and API examples on [GitHub](https://github.com/beancount/smart_importer?utm_source=chatgpt.com). ([GitHub][1])
* **PyPI smart-importer page** — installation and hooks description. ([[PyPI](https://pypi.org/project/smart-importer/?utm_source=chatgpt.com)][7])
* **Beancount advanced thread / mailing list** — usage discussion, training data hints, and workflow. ([[Mustachian Post Community](https://forum.mustachianpost.com/t/beancount-advancement-thread/2786?page=7&utm_source=chatgpt.com)][4])
* **Beancount import docs** — contextual information on how Beancount importers work generally. ([[beancount](https://beancount.github.io/docs/importing_external_data.html?utm_source=chatgpt.com)][2])
* **Community discussions** — practical user insights about prediction reliability and workflow. ([[Google Groups](https://groups.google.com/g/beancount/c/rjrbf6Y39ew?utm_source=chatgpt.com)][3])

---

If you’d like, I can produce a **ready-to-run example config + script** for a particular CSV format in your use case. Just tell me the bank/format you have.

[1]: https://github.com/beancount/smart_importer?utm_source=chatgpt.com "beancount/smart_importer"
[2]: https://beancount.github.io/docs/importing_external_data.html?utm_source=chatgpt.com "Importing External Data - Beancount Documentation"
[3]: https://groups.google.com/g/beancount/c/rjrbf6Y39ew?utm_source=chatgpt.com "smart importer newbie question"
[4]: https://forum.mustachianpost.com/t/beancount-advancement-thread/2786?page=7&utm_source=chatgpt.com "Beancount advancement Thread - Page 7 - Life Hacking"
[5]: https://blog.csdn.net/gitblog_00015/article/details/141744523?utm_source=chatgpt.com "Beancount Smart Importer 使用与安装教程原创"
[6]: https://github.com/bratekarate/beancount-categorizer?utm_source=chatgpt.com "bratekarate/beancount-categorizer: Decorator for ..."
[7]: https://pypi.org/project/smart-importer/?utm_source=chatgpt.com "smart-importer"
